tests/dataset_runner.py

```diff
@@ -0,0 +1,180 @@
+        self._run_command(['hca', 'upload', 'select', self.upload_credentials])
+        for file_path in self.dataset.data_files_paths():
+            self._run_command(['hca', 'upload', 'file', file_path])
+        self.forget_about_upload_area()
+
+    def _stage_data_files_using_s3_sync(self):
+        Progress.report("STAGING FILES using aws s3 sync...")
+        upload_area_uuid = self.upload_credentials.split(':')[4]
+        upload_area_s3_location = f"s3://org-humancellatlas-upload-{self.deployment}/{upload_area_uuid}"
+        command = ['aws', 's3', 'sync', '--content-type', 'application/gzip; dcp-type=data',
+                   self.dataset.config['data_files_location'],
+                   upload_area_s3_location]
+        self._run_command(command)
+
+    def forget_about_upload_area(self):
+        upload_area_uuid = self.upload_credentials.split(':')[4]
+        self._run_command(['hca', 'upload', 'forget', upload_area_uuid])
+
+    def wait_for_envelope_to_be_validated(self):
+        Progress.report("WAIT FOR VALIDATION...")
+        WaitFor(self._envelope_is_valid).to_return_value(value=True, timeout_seconds=15 * MINUTE)
+        Progress.report("  envelope is valid.\n")
+
+    def _envelope_is_valid(self):
+        envelope_status = self.submission_envelope.reload().status()
+        Progress.report(f"  envelope status is {envelope_status}")
+        return envelope_status in ['Valid', 'Submitted']
+
+    def complete_submission(self):
+        Progress.report("COMPLETING SUBMISSION...")
+        submit_url = self.submission_envelope.data['_links']['submit']['href']
+        response = requests.put(submit_url, headers=self.ingest_api.auth_headers)
+        if response.status_code != requests.codes.accepted:
+            raise RuntimeError(f"PUT {submit_url} returned {response.status_code}: {response.content}")
+        Progress.report("  done.\n")
+
+    def wait_for_primary_bundles_to_be_created(self):
+        Progress.report("WAITING FOR PRIMARY BUNDLE(s) TO BE CREATED...")
+        expected_primary_bundle_count = self._how_many_primary_bundles_do_we_expect()
+        Progress.report(f"  expecting {expected_primary_bundle_count} primary bundles")
+        try:
+            WaitFor(
+                self._primary_bundle_count
+            ).to_return_value(value=expected_primary_bundle_count, timeout_seconds=15 * MINUTE)
+        except TimedOut:
+            Progress.report("  We did not get all the primary bundles we expected, "
+                            "but let us continue and see how many results bundles we get.")
+            self.failure_reason = f"Only received {self._primary_bundle_count()} " \
+                                  f"of {expected_primary_bundle_count} primary bundles"
+        bundle_uuids = self.submission_envelope.bundles()
+
+        Progress.report(f"  {bundle_uuids}\n")
+        for bundle_uuid in bundle_uuids:
+            self.primary_to_results_bundles_map[bundle_uuid] = None
+
```

Comment:
Between primary bundles and secondary bundles, 2 kinds of information could be helpful here:

- # of notifications received by Green/Lira
- How many workflows are running/failed

During Mint's retro, we have had some discussions about improving the visibility/transparency of GreenBox to dcp, once we figure out a better way to do that, we could add some functions here to make this test more helpful/trackable.

Reply:
👍